File: 215-kth-largest-element-in-an-array/215-kth-largest-element-in-an-array.java
class Solution {
    public int findKthLargest(int[] nums, int k) {
        k = nums.length - k;
        //quickselect
        return quickselect(nums, 0, nums.length - 1, k);
    }
    public int quickselect(int[] nums, int l, int r, int k) {
        //base case
        if(l == r) return nums[l];
        int index = partition(nums, l, r);
        if(index == k) return nums[k];
        else if(index < k) {
            return quickselect(nums, index + 1, r, k);
        } else {
            return quickselect(nums, l, index - 1, k);
        }
    }
    public int partition(int[] nums, int l, int r) {
        //pivot index is r
        int pivot = nums[r];
        int wall = l;
        for(int i = l; i < r; i++) {
            if(nums[i] < pivot) {
                swap(nums, i, wall);
                wall++;
            }
        }
        swap(nums, wall, r);
        return wall;
    }
    public void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: 215-kth-largest-element-in-an-array/215-kth-largest-element-in-an-array.py
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        k = len(nums) - k + 1
        def quickselect(nums, k):
            if len(nums) <= 1:
                return nums[0]
            pivot = nums[len(nums)//2]
            left = [x for x in nums if x < pivot]
            mid = [x for x in nums if x == pivot]
            right = [x for x in nums if x > pivot]
            
            nums, i, j = left + mid + right, len(left), len(left) + len(mid)
            if k < i:
                return quickselect(left, k)
            elif k > j:
                return quickselect(right, k - j)
            else:
                return nums[k]
        return quickselect(nums, k - 1)